File: src/main/java/net/coderextreme/data/BindingOperations.java
package net.coderextreme.data;
import org.web3d.x3d.jsail.*;
import org.web3d.x3d.jsail.CADGeometry.*;
import org.web3d.x3d.jsail.Core.*;
import org.web3d.x3d.jsail.CubeMapTexturing.*;
import org.web3d.x3d.jsail.DIS.*;
import org.web3d.x3d.jsail.EnvironmentalEffects.*;
import org.web3d.x3d.jsail.EnvironmentalSensor.*;
import org.web3d.x3d.jsail.EventUtilities.*;
import org.web3d.x3d.jsail.Followers.*;
import org.web3d.x3d.jsail.Geometry2D.*;
import org.web3d.x3d.jsail.Geometry3D.*;
import org.web3d.x3d.jsail.Geospatial.*;
import org.web3d.x3d.jsail.Grouping.*;
import org.web3d.x3d.jsail.HAnim.*;
import org.web3d.x3d.jsail.Interpolation.*;
import org.web3d.x3d.jsail.KeyDeviceSensor.*;
import org.web3d.x3d.jsail.Layering.*;
import org.web3d.x3d.jsail.Layout.*;
import org.web3d.x3d.jsail.Lighting.*;
import org.web3d.x3d.jsail.NURBS.*;
import org.web3d.x3d.jsail.Navigation.*;
import org.web3d.x3d.jsail.Networking.*;
import org.web3d.x3d.jsail.ParticleSystems.*;
import org.web3d.x3d.jsail.Picking.*;
import org.web3d.x3d.jsail.PointingDeviceSensor.*;
import org.web3d.x3d.jsail.Rendering.*;
import org.web3d.x3d.jsail.RigidBodyPhysics.*;
import org.web3d.x3d.jsail.Scripting.*;
import org.web3d.x3d.jsail.Shaders.*;
import org.web3d.x3d.jsail.Shape.*;
import org.web3d.x3d.jsail.Sound.*;
import org.web3d.x3d.jsail.Text.*;
import org.web3d.x3d.jsail.Texturing3D.*;
import org.web3d.x3d.jsail.Texturing.*;
import org.web3d.x3d.jsail.Time.*;
import org.web3d.x3d.jsail.VolumeRendering.*;
import org.web3d.x3d.jsail.fields.*;
public class BindingOperations {
  public static void main(String[] args) {
    ConfigurationProperties.setXsltEngine(ConfigurationProperties.XSLT_ENGINE_NATIVE_JAVA);
    ConfigurationProperties.setDeleteIntermediateFiles(false);
    X3D model = new BindingOperations().initialize();
    System.out.print(model.validationReport().trim());
    model.toFileJSON("../data/BindingOperations.new.json");
    }
    public X3D initialize() {
      X3D X3D0 =  new X3D().setProfile("Immersive").setVersion("3.3")
      .setHead(new head()
        .addMeta(new meta().setName("title").setContent("BindingOperations.x3d"))
        .addMeta(new meta().setName("description").setContent("Illustrate Viewpoint binding operations (in gory detail!) as described in Chapter 4 concepts. Scene design: a TimeSensor clock drives and IntegerSequencer for each t0/t1/etc. event, and a customized Script node sends bind/unbind events to the correct Viewpoint. Display the browser console to see occurrence of each event."))
        .addMeta(new meta().setName("creator").setContent("Don Brutzman"))
        .addMeta(new meta().setName("created").setContent("5 January 2008"))
        .addMeta(new meta().setName("modified").setContent("22 July 2013"))
        .addMeta(new meta().setName("reference").setContent("BindingOperations.console.txt"))
        .addMeta(new meta().setName("reference").setContent("BindingStackOperations.png"))
        .addMeta(new meta().setName("reference").setContent("X3D for Web Authors, Section 2.5.1, Figure 4.1"))
        .addMeta(new meta().setName("reference").setContent("http://X3dGraphics.com"))
        .addMeta(new meta().setName("reference").setContent("https://www.web3d.org/x3d/content/examples/X3dResources.html"))
        .addMeta(new meta().setName("rights").setContent("Copyright Don Brutzman and Leonard Daly 2007"))
        .addMeta(new meta().setName("subject").setContent("X3D book, X3D graphics, X3D-Edit, http://www.x3dGraphics.com"))
        .addMeta(new meta().setName("identifier").setContent("http://X3dGraphics.com/examples/X3dForWebAuthors/Chapter04ViewingNavigation/BindingOperations.x3d"))
        .addMeta(new meta().setName("generator").setContent("X3D-Edit 3.3, https://savage.nps.edu/X3D-Edit"))
        .addMeta(new meta().setName("license").setContent("../license.html")))
      .setScene(new Scene()
        .addChild(new Viewpoint().setDEF("View1").setCenterOfRotation(new float[] {-6f,0f,0f}).setDescription("Viewpoint 1").setPosition(new float[] {-6f,0f,5f}))
        .addChild(new Viewpoint().setDEF("View2").setCenterOfRotation(new float[] {-2f,0f,0f}).setDescription("Viewpoint 2").setPosition(new float[] {-2f,0f,5f}))
        .addChild(new Viewpoint().setDEF("View3").setCenterOfRotation(new float[] {2f,0f,0f}).setDescription("Viewpoint 3").setPosition(new float[] {2f,0f,5f}))
        .addChild(new Viewpoint().setDEF("View4").setCenterOfRotation(new float[] {6f,0f,0f}).setDescription("Viewpoint 4").setPosition(new float[] {6f,0f,5f}))
        .addComments("Script initialization ought to first bind view5 below.")
        .addChild(new Group()
          .addChild(new Transform().setDEF("Text1").setTranslation(new float[] {-6f,0f,0f})
            .addChild(new Shape("Text1")
              .setGeometry(new Text().setString(new MFString0().getArray())
                .setFontStyle(new FontStyle().setDEF("CenterJustify").setJustify(new MFString1().getArray())))
              .setAppearance(new Appearance()
                .setMaterial(new Material().setDiffuseColor(new float[] {1f,0f,0f})))))
          .addChild(new Transform().setDEF("Text2").setTranslation(new float[] {-2f,0f,0f})
            .addChild(new Shape("Text2")
              .setGeometry(new Text().setString(new MFString2().getArray())
                .setFontStyle(new FontStyle().setUSE("CenterJustify")))
              .setAppearance(new Appearance()
                .setMaterial(new Material().setDiffuseColor(new float[] {0f,1f,0f})))))
          .addChild(new Transform().setDEF("Text3").setTranslation(new float[] {2f,0f,0f})
            .addChild(new Shape("Text3")
              .setGeometry(new Text().setString(new MFString3().getArray())
                .setFontStyle(new FontStyle().setUSE("CenterJustify")))
              .setAppearance(new Appearance()
                .setMaterial(new Material().setDiffuseColor(new float[] {0f,0f,1f})))))
          .addChild(new Transform().setDEF("Text4").setTranslation(new float[] {6f,0f,0f})
            .addChild(new Shape("Text4")
              .setGeometry(new Text().setString(new MFString4().getArray())
                .setFontStyle(new FontStyle().setUSE("CenterJustify")))
              .setAppearance(new Appearance()
                .setMaterial(new Material())))))
        .addComments("The following advanced animation sequence uses nodes covered in Chapters 7, 8 and 9.")
        .addComments("It does not need to be studied in this chapter.")
        .addChild(new Transform().setTranslation(new float[] {0f,-3f,8f})
          .addComments("notice this next Viewpoint has been transformed with the text, so its position is relative. it is called view5 in the Script.")
          .addChild(new Viewpoint().setDEF("ClickToAnimateView").setDescription("Select animation sequence").setPosition(new float[] {0f,0f,7f}))
          .addChild(new Shape()
            .setGeometry(new Text().setString(new MFString5().getArray())
              .setFontStyle(new FontStyle().setJustify(new MFString6().getArray())))
            .setAppearance(new Appearance()
              .setMaterial(new Material().setDiffuseColor(new float[] {0.8f,0.4f,0f}))))
          .addChild(new Shape()
            .setGeometry(new Box().setSize(new float[] {7f,1f,0.02f}))
            .setAppearance(new Appearance()
              .setMaterial(new Material().setTransparency(1f))))
          .addChild(new TouchSensor().setDEF("TextTouchSensor").setDescription("Click to begin animating viewpoint selections"))
          .addChild(new TimeSensor().setDEF("Clock").setCycleInterval(10d))
          .addChild(new ROUTE().setFromField("touchTime").setFromNode("TextTouchSensor").setToField("set_startTime").setToNode("Clock"))
          .addChild(new IntegerSequencer().setDEF("TimingSequencer").setKey(new MFFloat7().getArray()).setKeyValue(new MFInt328().getArray()))
          .addChild(new ROUTE().setFromField("fraction_changed").setFromNode("Clock").setToField("set_fraction").setToNode("TimingSequencer"))
          .addChild(new Script().setDEF("BindingSequencerEngine")
            .addField(new field("BindingSequencerEngine").setType("SFInt32").setName("set_timeEvent").setAccessType(field.ACCESSTYPE_INPUTONLY))
            .addField(new field("BindingSequencerEngine").setType("SFBool").setName("bindView1").setAccessType(field.ACCESSTYPE_OUTPUTONLY))
            .addField(new field("BindingSequencerEngine").setType("SFBool").setName("bindView2").setAccessType(field.ACCESSTYPE_OUTPUTONLY))
            .addField(new field("BindingSequencerEngine").setType("SFBool").setName("bindView3").setAccessType(field.ACCESSTYPE_OUTPUTONLY))
            .addField(new field("BindingSequencerEngine").setType("SFBool").setName("bindView4").setAccessType(field.ACCESSTYPE_OUTPUTONLY))
            .addField(new field("BindingSequencerEngine").setType("SFBool").setName("bindView5").setAccessType(field.ACCESSTYPE_OUTPUTONLY))
            .addField(new field("BindingSequencerEngine").setType("SFBool").setName("view1Bound").setAccessType(field.ACCESSTYPE_INPUTONLY))
            .addField(new field("BindingSequencerEngine").setType("SFBool").setName("view2Bound").setAccessType(field.ACCESSTYPE_INPUTONLY))
            .addField(new field("BindingSequencerEngine").setType("SFBool").setName("view3Bound").setAccessType(field.ACCESSTYPE_INPUTONLY))
            .addField(new field("BindingSequencerEngine").setType("SFBool").setName("view4Bound").setAccessType(field.ACCESSTYPE_INPUTONLY))
            .addField(new field("BindingSequencerEngine").setType("SFInt32").setName("priorInputvalue").setAccessType(field.ACCESSTYPE_INITIALIZEONLY).setValue("-1"))
            .setSourceCode("ecmascript:\n"+
"\n"+
"function initialize ()\n"+
"{\n"+
"    bindView5 = true;\n"+
"    Browser.print ('Timing script initialized and ready for activation');\n"+
"}\n"+
"function set_timeEvent (inputValue)\n"+
"{\n"+
"    if (inputValue == priorInputvalue)\n"+
"    {\n"+
"        return; // ignore repeated inputs\n"+
"    }\n"+
"    // new value provided\n"+
"    priorInputvalue = inputValue;\n"+
"    // Browser.print ('\\ntimeEvent inputValue=' + inputValue);\n"+
"\n"+
"    // mimics user execution of Figure 4.1 steps t_0 through t_8\n"+
"    if (inputValue == 0)\n"+
"    {\n"+
"        Browser.print ('\\n===========\\n time t0');\n"+
"        bindView1 = true;\n"+
"    }\n"+
"    else if (inputValue == 1)\n"+
"    {\n"+
"        Browser.print ('\\n===========\\n time t1');\n"+
"        bindView2 = true;\n"+
"    }\n"+
"    else if (inputValue == 2)\n"+
"    {\n"+
"        Browser.print ('\\n===========\\n time t2');\n"+
"        bindView3 = true;\n"+
"    }\n"+
"    else if (inputValue == 3)\n"+
"    {\n"+
"        Browser.print ('\\n===========\\n time t3');\n"+
"        bindView3 = false;\n"+
"    }\n"+
"    else if (inputValue == 4)\n"+
"    {\n"+
"        Browser.print ('\\n===========\\n time t4');\n"+
"        bindView1 = true;\n"+
"    }\n"+
"    else if (inputValue == 5)\n"+
"    {\n"+
"        Browser.print ('\\n===========\\n time t5');\n"+
"        bindView2 = false;\n"+
"    }\n"+
"    else if (inputValue == 6)\n"+
"    {\n"+
"        Browser.print ('\\n===========\\n time t6');\n"+
"        bindView1 = false;\n"+
"    }\n"+
"    else if (inputValue == 7)\n"+
"    {\n"+
"        Browser.print ('\\n===========\\n time t7');\n"+
"        bindView4 = true;\n"+
"\n"+
"    }\n"+
"    else if (inputValue == 8)\n"+
"    {\n"+
"        Browser.print ('\\n===========\\n time t8');\n"+
"        Browser.print (', no action, all done');\n"+
"        Browser.print ('\\n\\n');\n"+
"    }\n"+
"}\n"+
"\n"+
"function view1Bound (inputValue)\n"+
"{\n"+
"    Browser.print (', view1Bound ' + (inputValue));\n"+
"    if (priorInputvalue == -1) Browser.print ('\\n');\n"+
"}\n"+
"function view2Bound (inputValue)\n"+
"{\n"+
"    Browser.print (', view2Bound ' + (inputValue));\n"+
"}\n"+
"function view3Bound (inputValue)\n"+
"{\n"+
"    Browser.print (', view3Bound ' + (inputValue));\n"+
"}\n"+
"function view4Bound (inputValue)\n"+
"{\n"+
"    Browser.print (', view4Bound ' + (inputValue));\n"+
"}\n"+
"function view5Bound (inputValue)\n"+
"{\n"+
"    Browser.print (', view5Bound ' + (inputValue));\n"+
"}"))
          .addComments("drive Script with TimeSensor clock")
          .addChild(new ROUTE().setFromField("value_changed").setFromNode("TimingSequencer").setToField("set_timeEvent").setToNode("BindingSequencerEngine"))
          .addComments("Script will bind and unbind Viewpoint nodes")
          .addChild(new ROUTE().setFromField("bindView1").setFromNode("BindingSequencerEngine").setToField("set_bind").setToNode("View1"))
          .addChild(new ROUTE().setFromField("bindView2").setFromNode("BindingSequencerEngine").setToField("set_bind").setToNode("View2"))
          .addChild(new ROUTE().setFromField("bindView3").setFromNode("BindingSequencerEngine").setToField("set_bind").setToNode("View3"))
          .addChild(new ROUTE().setFromField("bindView4").setFromNode("BindingSequencerEngine").setToField("set_bind").setToNode("View4"))
          .addChild(new ROUTE().setFromField("bindView5").setFromNode("BindingSequencerEngine").setToField("set_bind").setToNode("ClickToAnimateView"))
          .addComments("Viewpoint nodes report bind and unbind events")
          .addChild(new ROUTE().setFromField("isBound").setFromNode("View1").setToField("view1Bound").setToNode("BindingSequencerEngine"))
          .addChild(new ROUTE().setFromField("isBound").setFromNode("View2").setToField("view2Bound").setToNode("BindingSequencerEngine"))
          .addChild(new ROUTE().setFromField("isBound").setFromNode("View3").setToField("view3Bound").setToNode("BindingSequencerEngine"))
          .addChild(new ROUTE().setFromField("isBound").setFromNode("View4").setToField("view4Bound").setToNode("BindingSequencerEngine"))))      ;
    return X3D0;
    }
protected class MFString0 {
  protected org.web3d.x3d.jsail.fields.MFString getArray() {
    return new org.web3d.x3d.jsail.fields.MFString(new java.lang.String[] {"View","# 1"});
  }
}
protected class MFString1 {
  protected org.web3d.x3d.jsail.fields.MFString getArray() {
    return new org.web3d.x3d.jsail.fields.MFString(new java.lang.String[] {"MIDDLE","MIDDLE"});
  }
}
protected class MFString2 {
  protected org.web3d.x3d.jsail.fields.MFString getArray() {
    return new org.web3d.x3d.jsail.fields.MFString(new java.lang.String[] {"View","# 2"});
  }
}
protected class MFString3 {
  protected org.web3d.x3d.jsail.fields.MFString getArray() {
    return new org.web3d.x3d.jsail.fields.MFString(new java.lang.String[] {"View","# 3"});
  }
}
protected class MFString4 {
  protected org.web3d.x3d.jsail.fields.MFString getArray() {
    return new org.web3d.x3d.jsail.fields.MFString(new java.lang.String[] {"View","# 4"});
  }
}
protected class MFString5 {
  protected org.web3d.x3d.jsail.fields.MFString getArray() {
    return new org.web3d.x3d.jsail.fields.MFString(new java.lang.String[] {"Click here to animate"});
  }
}
protected class MFString6 {
  protected org.web3d.x3d.jsail.fields.MFString getArray() {
    return new org.web3d.x3d.jsail.fields.MFString(new java.lang.String[] {"MIDDLE","BEGIN"});
  }
}
protected class MFFloat7 {
  protected org.web3d.x3d.jsail.fields.MFFloat getArray() {
    return new org.web3d.x3d.jsail.fields.MFFloat(new float[] {0f,0.1f,0.2f,0.3f,0.4f,0.5f,0.6f,0.7f,0.8f,1f});
  }
}
protected class MFInt328 {
  protected org.web3d.x3d.jsail.fields.MFInt32 getArray() {
    return new org.web3d.x3d.jsail.fields.MFInt32(new int[] {0,1,2,3,4,5,6,7,8,10});
  }
}
}

File: src/main/dart/net/coderextreme/data/BindingOperations.dart
// Copyright 2021 by John Carlson.  All rights reserved.
import 'dart:io';
import '../x3d.dart';
var X3D0 =  X3D(

      profile_ : SFString('Immersive'),
      version_ : SFString('3.3'),
      head_ : 
        head(
          meta_ : [
            meta(
              name_ : SFString('title'),
              content_ : SFString('BindingOperations.x3d')),

            meta(
              name_ : SFString('description'),
              content_ : SFString('Illustrate Viewpoint binding operations (in gory detail!) as described in Chapter 4 concepts. Scene design: a TimeSensor clock drives and IntegerSequencer for each t0/t1/etc. event, and a customized Script node sends bind/unbind events to the correct Viewpoint. Display the browser console to see occurrence of each event.')),

            meta(
              name_ : SFString('creator'),
              content_ : SFString('Don Brutzman')),

            meta(
              name_ : SFString('created'),
              content_ : SFString('5 January 2008')),

            meta(
              name_ : SFString('modified'),
              content_ : SFString('22 July 2013')),

            meta(
              name_ : SFString('reference'),
              content_ : SFString('BindingOperations.console.txt')),

            meta(
              name_ : SFString('reference'),
              content_ : SFString('BindingStackOperations.png')),

            meta(
              name_ : SFString('reference'),
              content_ : SFString('X3D for Web Authors, Section 2.5.1, Figure 4.1')),

            meta(
              name_ : SFString('reference'),
              content_ : SFString('http://X3dGraphics.com')),

            meta(
              name_ : SFString('reference'),
              content_ : SFString('https://www.web3d.org/x3d/content/examples/X3dResources.html')),

            meta(
              name_ : SFString('rights'),
              content_ : SFString('Copyright Don Brutzman and Leonard Daly 2007')),

            meta(
              name_ : SFString('subject'),
              content_ : SFString('X3D book, X3D graphics, X3D-Edit, http://www.x3dGraphics.com')),

            meta(
              name_ : SFString('identifier'),
              content_ : SFString('http://X3dGraphics.com/examples/X3dForWebAuthors/Chapter04ViewingNavigation/BindingOperations.x3d')),

            meta(
              name_ : SFString('generator'),
              content_ : SFString('X3D-Edit 3.3, https://savage.nps.edu/X3D-Edit')),

            meta(
              name_ : SFString('license'),
              content_ : SFString('../license.html'))]),
      Scene_ : 
        Scene(
          children_ : [
            Viewpoint(
              DEF_ : SFString('View1'),
              centerOfRotation_ : SFVec3f([SFDouble(-6), SFDouble(0), SFDouble(0)]),
              description_ : SFString('Viewpoint 1'),
              position_ : SFVec3f([SFDouble(-6), SFDouble(0), SFDouble(5)])),

            Viewpoint(
              DEF_ : SFString('View2'),
              centerOfRotation_ : SFVec3f([SFDouble(-2), SFDouble(0), SFDouble(0)]),
              description_ : SFString('Viewpoint 2'),
              position_ : SFVec3f([SFDouble(-2), SFDouble(0), SFDouble(5)])),

            Viewpoint(
              DEF_ : SFString('View3'),
              centerOfRotation_ : SFVec3f([SFDouble(2), SFDouble(0), SFDouble(0)]),
              description_ : SFString('Viewpoint 3'),
              position_ : SFVec3f([SFDouble(2), SFDouble(0), SFDouble(5)])),

            Viewpoint(
              DEF_ : SFString('View4'),
              centerOfRotation_ : SFVec3f([SFDouble(6), SFDouble(0), SFDouble(0)]),
              description_ : SFString('Viewpoint 4'),
              position_ : SFVec3f([SFDouble(6), SFDouble(0), SFDouble(5)])),
          /*Script initialization ought to first bind view5 below.*/

            Group(
              children_ : [
                Transform(
                  DEF_ : SFString('Text1'),
                  translation_ : SFVec3f([SFDouble(-6), SFDouble(0), SFDouble(0)]),
                  children_ : [
                    Shape(
                      geometry_ : 
                        Text(
                          string_ : MFString([SFString("View"), SFString("# 1")]),
                          fontStyle_ : 
                            FontStyle(
                              DEF_ : SFString('CenterJustify'),
                              justify_ : MFString([SFString("MIDDLE"), SFString("MIDDLE")]))),
                      appearance_ : 
                        Appearance(
                          material_ : 
                            Material(
                              diffuseColor_ : SFColor([SFDouble(1), SFDouble(0), SFDouble(0)]))))]),

                Transform(
                  DEF_ : SFString('Text2'),
                  translation_ : SFVec3f([SFDouble(-2), SFDouble(0), SFDouble(0)]),
                  children_ : [
                    Shape(
                      geometry_ : 
                        Text(
                          string_ : MFString([SFString("View"), SFString("# 2")]),
                          fontStyle_ : 
                            FontStyle(
                              USE_ : SFString('CenterJustify'))),
                      appearance_ : 
                        Appearance(
                          material_ : 
                            Material(
                              diffuseColor_ : SFColor([SFDouble(0), SFDouble(1), SFDouble(0)]))))]),

                Transform(
                  DEF_ : SFString('Text3'),
                  translation_ : SFVec3f([SFDouble(2), SFDouble(0), SFDouble(0)]),
                  children_ : [
                    Shape(
                      geometry_ : 
                        Text(
                          string_ : MFString([SFString("View"), SFString("# 3")]),
                          fontStyle_ : 
                            FontStyle(
                              USE_ : SFString('CenterJustify'))),
                      appearance_ : 
                        Appearance(
                          material_ : 
                            Material(
                              diffuseColor_ : SFColor([SFDouble(0), SFDouble(0), SFDouble(1)]))))]),

                Transform(
                  DEF_ : SFString('Text4'),
                  translation_ : SFVec3f([SFDouble(6), SFDouble(0), SFDouble(0)]),
                  children_ : [
                    Shape(
                      geometry_ : 
                        Text(
                          string_ : MFString([SFString("View"), SFString("# 4")]),
                          fontStyle_ : 
                            FontStyle(
                              USE_ : SFString('CenterJustify'))),
                      appearance_ : 
                        Appearance(
                          material_ : 
                            Material()))])]),
          /*The following advanced animation sequence uses nodes covered in Chapters 7, 8 and 9.*/
          /*It does not need to be studied in this chapter.*/

            Transform(
              translation_ : SFVec3f([SFDouble(0), SFDouble(-3), SFDouble(8)]),
              /*notice this next Viewpoint has been transformed with the text, so its position is relative. it is called view5 in the Script.*/
              children_ : [
                Viewpoint(
                  DEF_ : SFString('ClickToAnimateView'),
                  description_ : SFString('Select animation sequence'),
                  position_ : SFVec3f([SFDouble(0), SFDouble(0), SFDouble(7)])),

                Shape(
                  geometry_ : 
                    Text(
                      string_ : MFString([SFString("Click here to animate")]),
                      fontStyle_ : 
                        FontStyle(
                          justify_ : MFString([SFString("MIDDLE"), SFString("BEGIN")]))),
                  appearance_ : 
                    Appearance(
                      material_ : 
                        Material(
                          diffuseColor_ : SFColor([SFDouble(0.8), SFDouble(0.4), SFDouble(0)])))),

                Shape(
                  geometry_ : 
                    Box(
                      size_ : SFVec3f([SFDouble(7), SFDouble(1), SFDouble(0.02)])),
                  appearance_ : 
                    Appearance(
                      material_ : 
                        Material(
                          transparency_ : 1))),

                TouchSensor(
                  DEF_ : SFString('TextTouchSensor'),
                  description_ : SFString('Click to begin animating viewpoint selections')),

                TimeSensor(
                  DEF_ : SFString('Clock'),
                  cycleInterval_ : 10),

                ROUTE(
                  fromField_ : SFString('touchTime'),
                  fromNode_ : SFString('TextTouchSensor'),
                  toField_ : SFString('set_startTime'),
                  toNode_ : SFString('Clock')),

                IntegerSequencer(
                  DEF_ : SFString('TimingSequencer'),
                  key_ : MFFloat([SFFloat(0), SFFloat(0.1), SFFloat(0.2), SFFloat(0.3), SFFloat(0.4), SFFloat(0.5), SFFloat(0.6), SFFloat(0.7), SFFloat(0.8), SFFloat(1)]),
                  keyValue_ : MFInt32([SFInt32(0), SFInt32(1), SFInt32(2), SFInt32(3), SFInt32(4), SFInt32(5), SFInt32(6), SFInt32(7), SFInt32(8), SFInt32(10)])),

                ROUTE(
                  fromField_ : SFString('fraction_changed'),
                  fromNode_ : SFString('Clock'),
                  toField_ : SFString('set_fraction'),
                  toNode_ : SFString('TimingSequencer')),

                Script(
                  DEF_ : SFString('BindingSequencerEngine'),
                  field_ : [
                    field(
                      type_ : SFString("SFInt32"),
                      name_ : SFString('set_timeEvent'),
                      accessType_ : SFString("inputOnly")),

                    field(
                      type_ : SFString("SFBool"),
                      name_ : SFString('bindView1'),
                      accessType_ : SFString("outputOnly")),

                    field(
                      type_ : SFString("SFBool"),
                      name_ : SFString('bindView2'),
                      accessType_ : SFString("outputOnly")),

                    field(
                      type_ : SFString("SFBool"),
                      name_ : SFString('bindView3'),
                      accessType_ : SFString("outputOnly")),

                    field(
                      type_ : SFString("SFBool"),
                      name_ : SFString('bindView4'),
                      accessType_ : SFString("outputOnly")),

                    field(
                      type_ : SFString("SFBool"),
                      name_ : SFString('bindView5'),
                      accessType_ : SFString("outputOnly")),

                    field(
                      type_ : SFString("SFBool"),
                      name_ : SFString('view1Bound'),
                      accessType_ : SFString("inputOnly")),

                    field(
                      type_ : SFString("SFBool"),
                      name_ : SFString('view2Bound'),
                      accessType_ : SFString("inputOnly")),

                    field(
                      type_ : SFString("SFBool"),
                      name_ : SFString('view3Bound'),
                      accessType_ : SFString("inputOnly")),

                    field(
                      type_ : SFString("SFBool"),
                      name_ : SFString('view4Bound'),
                      accessType_ : SFString("inputOnly")),

                    field(
                      type_ : SFString("SFInt32"),
                      name_ : SFString('priorInputvalue'),
                      accessType_ : SFString("initializeOnly"),
                      value_ : SFString('-1')),
                  ],
ecmascript:eval (0

, function initialize ()
{
    bindView5 = true;
    Browser.print ('Timing script initialized and ready for activation');
}
, function set_timeEvent (inputValue)
{
    if (inputValue == priorInputvalue)
    {
        return; // ignore repeated inputs
    }
    // new value provided
    priorInputvalue = inputValue;
    // Browser.print ('\ntimeEvent inputValue=' + inputValue);

    // mimics user execution of Figure 4.1 steps t_0 through t_8
    if (inputValue == 0)
    {
        Browser.print ('\n===========\n time t0');
        bindView1 = true;
    }
    else if (inputValue == 1)
    {
        Browser.print ('\n===========\n time t1');
        bindView2 = true;
    }
    else if (inputValue == 2)
    {
        Browser.print ('\n===========\n time t2');
        bindView3 = true;
    }
    else if (inputValue == 3)
    {
        Browser.print ('\n===========\n time t3');
        bindView3 = false;
    }
    else if (inputValue == 4)
    {
        Browser.print ('\n===========\n time t4');
        bindView1 = true;
    }
    else if (inputValue == 5)
    {
        Browser.print ('\n===========\n time t5');
        bindView2 = false;
    }
    else if (inputValue == 6)
    {
        Browser.print ('\n===========\n time t6');
        bindView1 = false;
    }
    else if (inputValue == 7)
    {
        Browser.print ('\n===========\n time t7');
        bindView4 = true;

    }
    else if (inputValue == 8)
    {
        Browser.print ('\n===========\n time t8');
        Browser.print (', no action, all done');
        Browser.print ('\n\n');
    }
}

, function view1Bound (inputValue)
{
    Browser.print (', view1Bound ' + (inputValue));
    if (priorInputvalue == -1) Browser.print ('\n');
}
, function view2Bound (inputValue)
{
    Browser.print (', view2Bound ' + (inputValue));
}
, function view3Bound (inputValue)
{
    Browser.print (', view3Bound ' + (inputValue));
}
, function view4Bound (inputValue)
{
    Browser.print (', view4Bound ' + (inputValue));
}
, function view5Bound (inputValue)
{
    Browser.print (', view5Bound ' + (inputValue));
})),
              /*drive Script with TimeSensor clock*/

                ROUTE(
                  fromField_ : SFString('value_changed'),
                  fromNode_ : SFString('TimingSequencer'),
                  toField_ : SFString('set_timeEvent'),
                  toNode_ : SFString('BindingSequencerEngine')),
              /*Script will bind and unbind Viewpoint nodes*/

                ROUTE(
                  fromField_ : SFString('bindView1'),
                  fromNode_ : SFString('BindingSequencerEngine'),
                  toField_ : SFString('set_bind'),
                  toNode_ : SFString('View1')),

                ROUTE(
                  fromField_ : SFString('bindView2'),
                  fromNode_ : SFString('BindingSequencerEngine'),
                  toField_ : SFString('set_bind'),
                  toNode_ : SFString('View2')),

                ROUTE(
                  fromField_ : SFString('bindView3'),
                  fromNode_ : SFString('BindingSequencerEngine'),
                  toField_ : SFString('set_bind'),
                  toNode_ : SFString('View3')),

                ROUTE(
                  fromField_ : SFString('bindView4'),
                  fromNode_ : SFString('BindingSequencerEngine'),
                  toField_ : SFString('set_bind'),
                  toNode_ : SFString('View4')),

                ROUTE(
                  fromField_ : SFString('bindView5'),
                  fromNode_ : SFString('BindingSequencerEngine'),
                  toField_ : SFString('set_bind'),
                  toNode_ : SFString('ClickToAnimateView')),
              /*Viewpoint nodes report bind and unbind events*/

                ROUTE(
                  fromField_ : SFString('isBound'),
                  fromNode_ : SFString('View1'),
                  toField_ : SFString('view1Bound'),
                  toNode_ : SFString('BindingSequencerEngine')),

                ROUTE(
                  fromField_ : SFString('isBound'),
                  fromNode_ : SFString('View2'),
                  toField_ : SFString('view2Bound'),
                  toNode_ : SFString('BindingSequencerEngine')),

                ROUTE(
                  fromField_ : SFString('isBound'),
                  fromNode_ : SFString('View3'),
                  toField_ : SFString('view3Bound'),
                  toNode_ : SFString('BindingSequencerEngine')),

                ROUTE(
                  fromField_ : SFString('isBound'),
                  fromNode_ : SFString('View4'),
                  toField_ : SFString('view4Bound'),
                  toNode_ : SFString('BindingSequencerEngine'))])]));
void main() { exit(0); }
